### ISSUE TITLE
Endpoints returning 502 on every other request when using Kustomize

### ISSUE THREAD

**Issue by jaunruh** [kubernetes-sigs/kustomize#3867]
I am not sure if this is a bug, expected behaviour or if I am doing something wrong.

When using Kustomize in combination with [ingress-nginx](https://kubernetes.github.io/ingress-nginx/) and [cert-manager](https://cert-manager.io/docs/tutorials/acme/ingress/) while following the tutorial in the later link I stumble upon a weird error.

Steps taken:
- created a fresh kubernetes cluster
- created yaml files with Kustomize overlays for
- applied **the overlay**
- tested the setup with a [staging issuer](https://cert-manager.io/docs/tutorials/acme/ingress/#step-6-configure-let-s-encrypt-issuer)
- **everything working as expected**
- replaced the staging issuer with the production issuer
- sending requests to one of the endpoints results in a 502 on every second request -> non deterministic endpoint behaviour

I do not get the same behaviour when I copy the overlay files in to the base folder and apply the base directly without the `-k` flag.

**Comment by Shell32-Natsu** [comment]
Could you compare the result from kustomize between your original method and copying overlays?

**Comment by jaunruh** [comment]
Using `/usr/bin/diff` on the generated files generated by `kubectl kustomize ./ >> (base|prod).yaml` the only differences given are related to `annotations` and `variant`.

I just realised, that I use an `Ingress` object in the `prod` environment that is identical to the base one. So I am overwriting the Ingress with an identical one. Only in the `dev` environment I would actually need to replace the `Ingress` settings. Maybe this is the root of my error?

**Comment by Shell32-Natsu** [comment]
It's worth a try.

**Comment by jaunruh** [comment]
```bash
➜  prod git:(master) kubectl describe ingress main
Warning: extensions/v1beta1 Ingress is deprecated in v1.14+, unavailable in v1.22+; use networking.k8s.io/v1 Ingress
Name:             main
Namespace:        myapp
Address:          20.79.81.30
Default backend:  default-http-backend:80 (<error: endpoints "default-http-backend" not found>)
TLS:
  myapp-cert terminates one.domain.my, two.domain.my, three.domain.my
Rules:
  Host           Path  Backends
  ----           ----  --------
  one.domain.my
                 /   service-one:8080 (10.0.1.24:8080,10.0.1.38:8080,10.0.1.58:8080 + 1 more...)
  two.domain.my
                 /   service-two:3000 (10.0.1.24:3000,10.0.1.38:3000,10.0.1.58:3000 + 1 more...)
  three.domain.my
                 /   service-three:80 (10.0.1.24:80,10.0.1.38:80,10.0.1.58:80 + 1 more...)
Annotations:     cert-manager.io/issuer: letsencrypt-prod
                 kubernetes.io/ingress.class: nginx
                 nginx.ingress.kubernetes.io/cors-allow-credentials: true
                 nginx.ingress.kubernetes.io/cors-allow-headers: X-Request-ID, Authorization, Content-Type
                 nginx.ingress.kubernetes.io/cors-allow-methods: PUT, GET, POST, OPTIONS
                 nginx.ingress.kubernetes.io/cors-allow-origin: *
                 nginx.ingress.kubernetes.io/enable-cors: true
                 nginx.ingress.kubernetes.io/force-ssl-redirect: true
                 note: prod overlay
Events:          <none>
➜  prod git:(master) kubectl get pods
NAME                                       READY   STATUS    RESTARTS   AGE
service-two-84fb45fbd-drt7d                      1/1     Running   0          2d1h
service-two-7696ddb658-fl2xm   1/1     Running   0          2d1h
service-one-5c4957d875-6rwcg    1/1     Running   0          2d1h
postgres-64656fcc6b-cjcqn                  1/1     Running   0          2d1h
```
As one can see there seem to be multiple backends for each service, even tho each service is only running one pod. This does not happen using plain `kubectl`.

Describing one of the pods gives the first internal ip of the corresponding service:
```bash
➜  prod git:(master) kubectl describe pod service-one-5c4957d875-6rwcg
Name:         service-one-5c4957d875-6rwcg
Namespace:    myapp
Priority:     0
Node:         aks-exnodepool-38331632-vmss000000/10.0.1.4
Start Time:   Mon, 03 May 2021 10:20:21 +0200
Labels:       app=myapp
              pod-template-hash=5c4957d875
              variant=prod
Annotations:  note: prod overlay
Status:       Running
IP:           10.0.1.24
IPs:
  IP:           10.0.1.24
Controlled By:  ReplicaSet/service-one-5c4957d875
Containers:
  myapp-go:
    Container ID:   containerd://185676a9e680b38d94b32fe66072ea94193c76861b1b734d14f3740c330ac970
    Image:          <container image>
    Image ID:       <container image id>
    Port:           8080/TCP
    Host Port:      0/TCP
    State:          Running
      Started:      Mon, 03 May 2021 10:20:31 +0200
    Ready:          True
    Restart Count:  0
    Environment:
      SOME_ENV_VARS: "ENV_VAL"
    Mounts:
      ...
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  ....
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:          <none>
```

I would try to prevent reapplying the yaml files because I need to prevent the cert-manager from re-retrieving the tls cert. I am currently constantly running into the letsencrypt rate limits.

**Comment by jaunruh** [comment]
Any updates on this? I am still having this issue and its very counter productive to our infrastructure.

**Comment by Shell32-Natsu** [comment]
@monopole Can you take a look?

**Comment by jaunruh** [comment]
I am still having this issue. I am not entirely sure if it is Kustomize related. It seems like either a general K8s issue or an nginx-ingress issue. 
I have encountered this issue again in the following scenario:
I used terraform to set up a `dev` environment. The `dev` environment works flawlessly.
We have a running `prod` environment that was setup manually which should be replaced. So I set up a new prod environment using the same terraform files as the ones used for the `dev` environment (therefore they should be identical).
Within the `prod` environment I used `kubectl` and `kustomize` to setup my containers for `staging` with `staging.*` subroutes in order to test if everything works properly. I then moved all the data from the old `prod` to the new (terraform) `prod`. Finally I only replaced the `ingress` with an ingress that used the normal non-prefix urls. So `staging.myurl.com` was replaced by `myurl.com`. After this step the random 502 responses returned. 
Everything in both environments ((new) `prod` and `dev`) is exactly the same now, apart from the urls and the fact, that the `dev` environment was never using a different url endpoint.
